AUS-213: Added CUPPA v2 output into ORANGE while maintaining compatibility for CUPPA v1

luan-n-nguyen opened the pull request:
Hi guys,

As we (me, Paul and Khalid) discussed, I've started a new branch and made new changes to have as clean of a pull request as possible for updating ORANGE with CUPPAv2 output. Below is a high level overview of the changes.

**orange** 
The orange pdf now only shows the CUPPA v2 plot.

I've also removed these values from the front page: max_complex_size, telomeric_sgl, number_of_line_insertions

**orange-datamodel**
CUPPA v2 and v1 probabilities now exist together in the same json format. I've added the field `cuppaMajorVersion` to distinguish old vs new CUPPA. 

Below is a snippet of `minimally.populated.orange.json`. I've manually edited this file as well as `minimally.empty.orange.json` and `real.orange.json` so that all the tests work. The CUPPA json entry now also contains the `bestPrediction` field.
```
"cuppa": {
    "bestPrediction": {
      "cuppaMajorVersion": "v1",
      "cancerType": "Melanoma",
      "likelihood": 0.996,
      "snv96Classifier": 0.979,
      "genomicPositionClassifier": 0.990,
      "eventClassifier": 0.972,
      "altSjClassifier": null,
      "geneExpressionClassifier": null
    },
    "predictions": [
      {
        "cuppaMajorVersion": "v1",
        "cancerType": "Melanoma",
        "likelihood": 0.996,
        "snv96Classifier": 0.979,
        "genomicPositionClassifier": 0.990,
        "eventClassifier": 0.972,
        "altSjClassifier": null,
        "geneExpressionClassifier": null
      }
    ]
```


I've tried to design the new orange-datamodel (as well as orange) code so that the old CUPPA v1 code is easy to remove once it is fully deprecated. For example CUPPA v1 probabilities are optional/nullable in `CuppaDataFactory`:
```
public final class CuppaDataFactory
{
    @NotNull
    public static CuppaData create(
            @NotNull CuppaPredictions cuppaPredictionsV2,
            @Nullable List<CuppaDataFile> cuppaPredictionsV1
    ){
```

kduyvesteyn reviewed (commented):
Will leave more detailed review to Khalid and Paul but have a few general questions:
 - My impression is that ideally we want to keep orange-datamodel backwards compatible but we are fine switching ORANGE itself to v2. I see now that ORANGE itself supports both v1 and v2, but I imagine upgrading ORANGE together with CUPPA is fine anyways (its not unusual to upgrade several tools in combination in pipeline to make sure they are compatible). I might have missed some details though from discussion between Luan and Paul so may not be entirely up to date. 
 - Should the major version of cuppa be on the level of an individual prediction? Feels more logical to have it on the top level (I imagine we don’t want to support mixed predictions from v1 and v2?), but not quite sure we should have it in the first place. The idea is that consumers of orange-datamodel don’t have to care about the cuppa version. If they would then it would be fine to make the change non-backwards-compatible anyways I feel. 
 - Why did you drop the traits that we extracted from cuppa (max complex size, number of line insertions, etc)? I appreciate them since they give a better understanding of what is going on and what is possible with WGS (these fields are similar to fields such as “whole genome duplication”, or the linx driver plots)?

luan-n-nguyen commented:
> * Why did you drop the traits that we extracted from cuppa (max complex size, number of line insertions, etc)

I can put these back into ORANGE.
I thought that these fields are no longer in the output of CUPPA v2, but turns out i did put them in.

luan-n-nguyen commented:
> * Should the major version of cuppa be on the level of an individual prediction?

Having the CUPPA json entry structured like `{bestPrediction: ..., predictionsV2: ..., predictionsV1: ...}` is indeed better, but i felt that required more changes to the codebase.

Once CUPPA v1 is deprecated, `cuppaMajorVersion` can simply be removed. We won't need to refactor any other code that separates v1 from v2.

pauldwolfe commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/cuppa/CuppaDataFactory.java:
```diff
@@ -26,19 +32,101 @@
+        CuppaPredictionEntry predictionEntry = cuppaPredictions.PredictionEntries.stream()
+                .filter(o -> o.DataType == Categories.DataType.FEAT_CONTRIB & o.FeatName.equals(featureName))
+                .findFirst()
+                .get();
```
this isn't great use of java Optional. It would be better to do get().orElseThrow() with an exception with a good message to the user about the problem in data integrity.

luan-n-nguyen replied:
`.get()` is now replaced with `.orElseThrow(() -> new Exception("Input CuppaPredictions is empty"))`

kzuberihmf commented on orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/cuppa/CuppaData.java:
```diff
@@ -17,12 +17,19 @@ public interface CuppaData
     @NotNull
     List<CuppaPrediction> predictions();
 
-    int simpleDups32To200B();
```
Oh, I see this was later added back in. I'm glad you found these were still available in v2 @luan-n-nguyen, having these again removes one of my worries for this upgrade.

Though is the change of type from int to Integer necessary? How would that affect compat?

luan-n-nguyen replied:
I made these Integers so that they are nullable so that `doNotCrashOnMissingEntries` passes in [CuppaDataFactoryTest](https://github.com/hartwigmedical/hmftools/blob/e4ec8441ae4f8e690639cd76e55185f9f1afcff1/orange/src/test/java/com/hartwig/hmftools/orange/algo/cuppa/CuppaDataFactoryTest.java#L143C19-L143C19)

kzuberihmf replied:
Could you clarify this a bit more? That test looks like it was originally added several months ago (though modified recently), and they couldn't be null previously. Does the cuppa v2 make it possible for these being null in some cases? If not, something seems suspicious about that test case.

pauldwolfe replied:
This change is not backwards compatible, while it's an improvement, old clients could now fail with an NPE as they'd expect the magic -1.  We need to revert it back to @NotNull/primitive.

luan-n-nguyen replied:
I've now made the SV feature values be primitive ints again, and have them be -1 if they are "null"

kzuberihmf commented on orange/src/main/java/com/hartwig/hmftools/orange/algo/OrangeAlgo.java:
```diff
@@ -573,22 +574,25 @@ private static CuppaData loadCuppaData(@NotNull OrangeConfig config) throws IOEx
-        CuppaPrediction best = cuppaData.predictions().get(0);
-        LOGGER.info(" Predicted cancer type '{}' with likelihood {}", best.cancerType(), best.likelihood());
+        List<CuppaDataFile> cuppaPredictionsV1 = null;
+        String cuppaResultTsv = config.wgsRefConfig().cuppaResultCsv();
```
Right, this was probably the confusing bit. We are ok with the orange application supporting only cuppa v2 going forward, but we want the *orange-datamodel* to be backwards compatible.  Tbh it wasn't even perfectly clear to me that was the goal earlier on, but glad we are converging on an understanding 😄.

luan-n-nguyen replied:
Ok! That makes things easier. I'll remove the support for CUPPA v1 in OrangeAlgo, OrangeApplication, etc

kzuberihmf commented on orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/cuppa/CuppaPrediction.java:
```diff
@@ -10,23 +10,26 @@
     double likelihood();
 
     @Nullable
-    Double snvPairwiseClassifier();
```
Is this just an inadvertent renaming of the fields? E.g. it looks like altSjCohortClassifier() became altSjClassifier(). All the original names would need to stay as Paul indicated, and preferably be populated by semantically equivalent values from v2.

luan-n-nguyen replied:
I renamed them as the old names were no longer technically correct for CUPPA v2.
However, for compatibility's sake i can revert to the old names.

kzuberihmf commented:
orange-datamodel in master still has cuppa2 code, which doesn't show up on this diff, but we decided we don't want that split. Happily it doesn't seem to be used anymore, so likely easy to remove.

So looks to me like (summarizing various changes):
* can drop reading of cuppa v1 in orange 
* drop the cuppaMajorVersion in CuppaPrediction. We acknowledge we might later discover clients actually want this for some reason, if so we'll add it it in then but at a higher level. As Korneel says we don't expect to support the case of a single orange-datamodel instance/json being populated by predictions from multiple cuppa versions
* put the v1 plot fields back in, just don't populate them, they are nullable and all current users should already be able to safely handle them being missing
* for the v2 plot, you have an option, either put it as a new field, or consider re-using cuppaSummaryPlot(). In the latter case, clients will likely just pull it in for display without requiring change I would think. Others can chime in if this doesn't make sense
* restore the names of those field that *I think* are the same but have been renamed. If there are genuinely new fields required, the old fields still must be put back for compatibility. 
* For double extra safety, consider doing a diff against the last revision that had orange-datamodel changes, to make sure that we don't miss anything that's still lost in master :smile:. I believe that would be against the full tree as of this commit 422052d39fa9eb373d46c621688dd00206d65ebf. You should be able to verify that all the orange-datamodel fields that existed then still exist, same names and types etc. New fields are ok where they add something new that's only in v2, though I think we felt earlier it was likely that v2 attributes could be mapped to the currently existing cuppa attributes in orange-datamodel

Hope that makes sense, and of course we can always sync up again in a meeting if preferred. I think this is already most of the way there, with just (hopefully!) some shallow renaming & cleanups left to do.

luan-n-nguyen commented on orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/orange/OrangePlots.java:
```diff
@@ -43,11 +43,5 @@ public interface OrangePlots
     List<String> linxDriverPlots();
 
     @Nullable
-    String cuppaSummaryPlot();
```
I'll use `cuppaSummaryPlot` to insert the CUPPA v2 visualization as you suggested below

luan-n-nguyen commented on hmf-common/src/test/java/com/hartwig/hmftools/common/cuppa2/CuppaPredictionsTest.java:
```diff
@@ -26,7 +26,7 @@ public void subsetByClfNameReturnsOneClfName()
 
         List<Categories.ClfName> clfNames = predictionEntries.stream()
-                .map(CuppaPredictionEntry::getClfName)
+                .map(o -> o.ClfName)
```
I use a lambda here because `CuppaPredictionEntry` does not have getters

pauldwolfe replied:
You can use a method reference irrespective of whether the method is a getter.

luan-n-nguyen commented:
Hi all,

I've now made the changes suggested by @kzuberihmf:

- CUPPA v1 is now no longer supported in orange. Old CUPPA properties such as `cuppaFeaturePlot`, `cuppaChartPlot` and `cuppaResultsCsv` still remain (or were restored) in the `OrangeWGSRefConfig` and/or `OrangePlots` immutable classes, but they are no longer called by `OrangeAlgo`.
- The CUPPA v2 plot is now stored in the `cuppaSummaryPlot` property
- `cuppaMajorVersion` is now removed from orange-datamodel
- Classifier names are restored to the old names: snv96Classifier -> snvPairwiseClassifier, eventClassifier -> featureClassifier, altSjClassifier -> altSjCohortClassifier, geneExpressionClassifier -> expressionPairwiseClassifier

kzuberihmf commented:
Thanks for the updates!

> 
> * CUPPA v1 is now no longer supported in orange. Old CUPPA properties such as `cuppaFeaturePlot`, `cuppaChartPlot` and `cuppaResultsCsv` still remain (or were restored) in the `OrangeWGSRefConfig` and/or `OrangePlots` immutable classes, but they are no longer called by `OrangeAlgo`.

Something suggested by @kduyvesteyn, since these fields need to be kept for compatibility (which we are doing) but will no longer be populated going forward, we can signal this to client code by marking them with an @Deprecated annotation. Tooling should then give provide hints about the change.

kzuberihmf commented:
Also, I still see cuppa2 in orange-datamodel in this branch, https://github.com/hartwigmedical/hmftools/tree/AUS-213/orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/cuppa2. Why is this here?

luan-n-nguyen commented:
> Also, I still see cuppa2 in orange-datamodel in this branch, https://github.com/hartwigmedical/hmftools/tree/AUS-213/orange-datamodel/src/main/java/com/hartwig/hmftools/datamodel/cuppa2. Why is this here?

I was intending to delete this but forgot. Thanks for catching this :). It's now deleted.

> by marking them with an @deprecated annotation

I've now added the @Deprecated tag to cuppaResultCsv, cuppaChartPlot, and cuppaFeaturePlot

pauldwolfe reviewed (approved):
Looks good, thanks @luan-n-nguyen
